### ISSUE TITLE
Profiled Media Type (MIME) for STAC dcouments

### ISSUE THREAD

**Issue by emmanuelmathot** [radiantearth/stac-spec#851]
# Summary

Need for a specific media type definition that specifies modeling level for Item, Catalog and Collection.

# Context

Exposing a Web Service (OGC API Processes) with STAC documents as input. The parameter declares the media type (MIME) of expected input. The current implementation uses `application/json` and `application/geo+json`. This syntactical level of specification is necessary but insufficient to indicate that the expected JSON document must follow STAC specification.

# Proposal

A parameter can be defined as additional semantics that can be used to process a resource representation, such as constraints, conventions, extensions, or any other aspects that do not alter the basic media type semantics, which covers our use cases of structures and models. Concretely, following https://tools.ietf.org/html/rfc2046 we would define a profile extension to the media-type to specify the profile of the JSON:
- `application/geo+json; profile=stac-item`
- `application/json; profile=stac-collection`
- `application/json; profile=stac-catalogue`

# Alternatives

Another option is to defined a vendor-specific mime-type such as "application/stac-item+json", "application/stac-collection+json", "application/stac-catalog+json"

Let me know the best practice or any eventual specifications
Thanks

**Comment by m-mohr** [comment]
Some remarks:
1. Implicitly, the media type for STAC Items is currently `application/geo+json` and for Catalogs/Collections is `application/json`. Introducing these new media types would have influence on existing STACs as the links would need a different media types in `type`. Alternative would be to allow new media types with and without profile.
2. This would also influence work on #848 likely other things that built on top of STAC.
3. Do implementations actually parse media types so that they may just ignore the profile if set? I doubt it.

**Comment by emmanuelmathot** [comment]
Some answers:

1. The proposal would not change the media type, it remais `application/geo+json` and `application/json`. Alternatives do but we prefer not doing that. `profile` is a parameter in the content-type header or media-type and is not mandatory.
2. Since the core media-type remains the same, there is no impact. The problem resides in the software implementation. See 3.
3. Indeed, it all depends on the software implementation. Most of the software libraries parses properly a mime-type separating the core media-type from the parameters. They must often handle `charset` or `boundary` parameter so it should not be a problem if the parsing follows convention.

So in summary yes, the proposal is just about fixing a common convention for the `profile` parameter if needed.

**Comment by m-mohr** [comment]
2. Have you read into the linked PR? Doesn't seem like that, because it impact on it (likely in a positive way).
3. Probably works with libraries where the core business is handling such things, but I have often seen code like `if(link.type == 'application/geo+json') ...` in applications, which would break...

**Comment by emmanuelmathot** [comment]
2. From what I saw, maybe too quickly, the PR #848 was about specifying a relationship type in the links pointing to a STAC document which is close but with a slightly different purpose.
Anyway, there is a synergy with our proposal that could help, as a best practice, to identify STAC documents, by specifying this `profile` parameter in the mime type, typically in the HTTP header `Content-Type` or in the links media-type. Our objective is to use this best practice a specification for STAC as input requirement for a process (e.g. OGC API processes)
3. I know this implementation issue but this kind of code will anyway cause troubles when the web server shall specify the `charset` for instance like `application/json; charset=utf-8`

**Comment by m-mohr** [comment]
3. I was more thinking about the media types in the STAC files (i.e. link `type`).

**Comment by emmanuelmathot** [comment]
2. Agreed for API but there could be also a convention specified in the core STAC spec.
3. The link `type` attribute can contain also the media-type parameters.

Let me show you the purpose in 2 practical examples in 2 different data model.

1. STAC data model in JSON encoding
```json
  "links": [
    {"rel": "self", "type": "application/json; profile=stac-item", "href": "http://cool-sat.com/catalog/CS3-20160503_132130_04/CS3-20160503_132130_04.json"},
    {"rel": "root", "type": "application/json; profile=stac-catalog", "href": "http://cool-sat.com/catalog/catalog.json"},
    {"rel": "collection", "type": "application/json; profile=stac-collection", "href": "http://cool-sat.com/catalog/CS3-20160503_132130_04/catalog.json"},
    {"rel": "alternate", "type": "text/html", "href": "http://cool-sat.com/catalog/CS3-20160503_132130_04/CS3-20160503_132130_04.html"}
  ]
```
in the above example, the profiles are quite useless because there are implicitly defined by the `rel` attribute in the STAC domain. Even though, knowing the difference between collection and catalog could be useful.

2. ATOM data model in XML encoding
```xml
<link rel="self" type="application/atom+xml" title="Reference link" href="https://catalog.terradue.com/sentinel2/search?format=atom&uid=S2A_MSIL1C_20190119T074231_N0207_R092_T39TVH_20190119T091330" />
<link rel="enclosure" type="application/octet-stream" title="Sentinels Scientific data hub zip file" href="https://scihub.copernicus.eu/apihub/odata/v1/Products('0eb80572-fc3c-4aeb-bd25-8c48796fbb77')/$value" />
<link rel="icon" type="image/jpg" title="Browse" href="https://store.terradue.com/api/scihub/sentinel2/S2MSI1C/2019/01/19/quicklooks/v1/S2A_MSIL1C_20190119T074231_N0207_R092_T39TVH_20190119T091330.jpg" />
<link rel="alternate" type="application/xml; profile=opengis/eop2.1" title="EOP profile" href="https://catalog.terradue.com//eop/sentinel2/profile/xml?uid=S2A_MSIL1C_20190119T074231_N0207_R092_T39TVH_20190119T091330" />
<link rel="self" type="application/geo+json" title="Reference link in JSON" href="https://catalog.terradue.com/sentinel2/search?format=json&uid=S2A_MSIL1C_20190119T074231_N0207_R092_T39TVH_20190119T091330" />
<link rel="alternate" type="application/geo+json; profile=stac-item" title="EOP profile" href="https://sentinel-stac.s3.amazonaws.com/sentinel-2-l1c/39/T/VH/2019-01-19/S2A_39TVH_20190119_0.json" />
<link rel="alternate" type="application/xml; profile=opengis/eop2.1" title="STAC Item" href="https://catalog.terradue.com//eop/sentinel2/profile/xml?uid=S2A_MSIL1C_20190119T074231_N0207_R092_T39TVH_20190119T091330" />
<link rel="alternate" type="application/json; profile=stac-collection" title="STAC collection" href="https://sentinel-stac.s3.amazonaws.com/sentinel-2-l1c/catalog.json" />
```

In that last example, I am in another data model. If I specify simply a link with a type in `application/geo+json`, I have no functional way to choose between the `Reference link in JSON` and the `STAC Item` links. With `application/geo+json; profile=stac-item`, the client knows it will find a stac item following the link.

**Comment by cholmes** [comment]
There's some additional discussion of this topic in #918

**Comment by cholmes** [comment]
Ok, I put up a PR to spur discussion. Let me know if it looks right. I do think a 'best practices' may be in order to add some more info about using this, and we should open an issue in STAC API spec as it's affected.

**Comment by m-mohr** [comment]
An important note from OGC that also affects our media type choice: https://github.com/opengeospatial/geotiff/issues/34#issuecomment-770897766

**Comment by m-mohr** [comment]
I think with that in mind we can't use the proposed media types and need to come up with our own, which are not adaptions from application/json and application/geo+json. Having that said, I'm afraid that we break tooling when servers will return new media types that are unknown. Like currently many HTTP libraries will automatically parse JSON due to the media type, but once servers return application/stac-item or so, that will probably not happen any more. Even browsers may offer you stac files for download then instead of just serving them nicely formatted as JSON. I think this may lead to more issues than it actually solves and thus my preference at the moment would be to stick with the JSON and GeoJSON media types. Maybe things impove a bit with the type field we add to Catalogs and Collections in #889. If we decide to have new media types, we should run intensive tests beforehand with a number of common implementations. This would delay RC1 (for weeks?), but on the same hand we should probably have it in for the RC phase.

**Comment by cholmes** [comment]
Yeah, I agree that we should stick with json / geojson types, and then drop our 'profile' stuff. It seems far more useful for clients to know that the linked files are json / geojson than and be able to do 'something' with it, vs 'application/vnd.stac-catalog' or whatever that most clients won't recognize. One idea would be to add some extra parameter to our 'link' object, so you'd say 'type=application/json' and then 'type-profile' or 'sub-type' or something equals 'stac-catalog'.

**Comment by m-mohr** [comment]
Most links should already indicate it by the rel type whether something is an item or catalog/collection. Not sure whether there's much benfit of an additional field.
